arbiter-core/src/environment/mod.rs

```diff
@@ -475,7 +482,7 @@ impl Environment {
                         // Set the tx_env and prepare to process it
                         evm.env.tx = tx_env;
 
-                        let result = evm.transact()?.result;
+                        let result = evm.transact_ref()?.result;
```

Comment:
Little confused on this, could you explain what happens here?

Reply:
Yes. There are 3 traits for the DB's in Revm: `Database`, `DatabaseCommit`, and  `DatabaseRef`.

All the methods for both `Database` and `DatabaseCommit` take in the db as exclusive reference`&mut` whereas the methods in `DatabaseRef` take in only as ref `&`.  The reason why we changed from `evm.transact()` to `evm.transact_ref()` is because  the implementation I used for `ArbiterDB` for these traits uses `self.0.write()` for the `&mut` and `self.0.read()` for `&` for the `DatabaseRef` version. This is done so that if the environment is currently handling a read-only instruction here, then the `RwLock` is not blocking other read-only instructions from being used on shared state.